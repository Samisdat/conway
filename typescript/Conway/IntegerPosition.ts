import {Position} from './Position';
import {Integer} from './Integer';

export class IntegerPosition extends Position {

    public readonly  x: number;
    public readonly  y: number;

    constructor(x: number, y: number) {
        super(x, y);

        this.x = new Integer(x).value;
        this.y = new Integer(y).value;

    }

    public move(position: IntegerPosition): IntegerPosition {
        return new IntegerPosition(
            this.x + position.x,
            this.y + position.y
        );
    }

    public inverse(): IntegerPosition {
        return new IntegerPosition(
            -1 * this.x,
            -1 * this.y
        );
    }

    public compare(position: IntegerPosition): Boolean {

        return (position.x === this.x && position.y === this.y);

    }

    public getNeighbours(): IntegerPosition[] {

        let neighbours: IntegerPosition[] = [];

        // neighbours on bottom
        neighbours.push(
            new IntegerPosition(this.x - 1, this.y - 1)
        );
        neighbours.push(
            new IntegerPosition(this.x, this.y - 1)
        );
        neighbours.push(
            new IntegerPosition(this.x + 1, this.y - 1)
        );

        // neighbours on the left and on the right
        neighbours.push(
            new IntegerPosition(this.x - 1, this.y)
        );
        neighbours.push(
            new IntegerPosition(this.x + 1, this.y)
        );

        // neighbours on top
        neighbours.push(
            new IntegerPosition(this.x - 1, this.y + 1)
        );
        neighbours.push(
            new IntegerPosition(this.x, this.y + 1)
        );
        neighbours.push(
            new IntegerPosition(this.x + 1, this.y + 1)
        );

        return neighbours;
    }

    public clone(): IntegerPosition {
        return new IntegerPosition(
            this.x,
            this.y
        );
    }

}
