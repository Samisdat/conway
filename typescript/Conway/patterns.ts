
// taken from
// http://en.wikipedia.org/wiki/Conway%27s_Game_of_Life
import {Pattern} from './pattern';

interface PatternsInterface {
    [index: string]: number[][];
}

const patterns: PatternsInterface = {
    // debug
    rotate: [
        [1, 0, 0, 0],
        [1, 0, 0, 0],
        [1, 0, 0, 0],
        [1, 1, 1, 1]
    ],

    // still lives
    block: [
        [1, 1],
        [1, 1]
    ],
    beehive: [
        [0, 1, 1, 0],
        [1, 0, 0, 1],
        [0, 1, 1, 0]
    ],
    loaf: [
        [0, 1, 1, 0],
        [1, 0, 0, 1],
        [0, 1, 0, 1],
        [0, 0, 1, 0]
    ],
    boat: [
        [1, 1, 0],
        [1, 0, 1],
        [0, 1, 0]
    ],
    // Oscillators
    blinker: [
        [0, 0, 0],
        [1, 1, 1],
        [0, 0, 0]
    ],
    toad: [
        [0, 1, 1, 1],
        [1, 1, 1, 0]
    ],
    beacon: [
        [1, 1, 0, 0],
        [1, 1, 0, 0],
        [0, 0, 1, 1],
        [0, 0, 1, 1]
    ],
    pulsar: [
        [0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1],
        [1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1],
        [1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1],
        [0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0],
        [1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1],
        [1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1],
        [1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    ],
    revolver: [
        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
        [1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1],
        [0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0],
        [0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0],
        [0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0],
        [0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0],
        [1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1],
        [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
    ],
    // Spaceships
    glider: [
        [0, 1, 0],
        [0, 0, 1],
        [1, 1, 1]
    ],
    lightweight_spaceship: [
        [0, 1, 0, 0, 1],
        [1, 0, 0, 0, 0],
        [1, 0, 0, 0, 1],
        [1, 1, 1, 1, 0]
    ],
    // Methuselahs
    r_pentomino: [
        [0, 1, 1],
        [1, 1, 0],
        [0, 1, 0]
    ],
    diehard: [
        [0, 0, 0, 0, 0, 0, 1, 0],
        [1, 1, 0, 0, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 1, 1, 1]
    ],
    acorn: [
        [0, 1, 0, 0, 0, 0, 0],
        [0, 0, 0, 1, 0, 0, 0],
        [1, 1, 0, 0, 1, 1, 1]
    ],
    gosper_glider_gun: [
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0],
        [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    ],
    exhibit_infinite_growth_1: [
        [0, 0, 0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 1, 0, 1, 1],
        [0, 0, 0, 0, 1, 0, 1, 0],
        [0, 0, 0, 0, 1, 0, 0, 0],
        [0, 0, 1, 0, 0, 0, 0, 0],
        [1, 0, 1, 0, 0, 0, 0, 0]
    ],
    glider_eater: [
        [0, 0, 0, 0, 0, 0, 0, 1, 1],
        [0, 0, 0, 1, 0, 0, 0, 1, 1],
        [0, 0, 1, 0, 1, 0, 0, 0, 0],
        [0, 1, 0, 1, 0, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0, 0, 0, 0],
        [1, 1, 0, 0, 0, 0, 0, 0, 0]
    ],
    guns_and_eaters: [
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
        [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0],
        [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1],
        [0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1],
        [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    ]
};

export class Patterns {

    public get(name: string): Pattern {

        if (undefined === patterns[name]) {
            throw new Error('unkown pattern ' + name);
        }

        return new Pattern(name, patterns[name]);

    }

}
