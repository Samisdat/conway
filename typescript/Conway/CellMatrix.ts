import {Cell} from './cell';
import {Position} from './position';

interface CellMatrixInterface {
    [index: string]: Cell;
}

export class CellMatrix {

    private matrix: CellMatrixInterface = {};

    public get(position: Position): Cell | undefined {

        if (undefined !== this.matrix[position.toString()]) {
            return this.matrix[position.toString()];
        }

        return undefined;


    }

    public add(cell: Cell): void {

        if (undefined !== this.matrix[cell.position.toString()]) {
            console.log('argh');
        }

        this.matrix[cell.position.toString()] = cell;

    }

    public remove(cell: Cell): void {

        if (undefined !== this.matrix[cell.position.toString()]) {
            delete this.matrix[cell.position.toString()];
        }

    }

    public all(): Cell[] {

        const cells: Cell[] = [];

        for (const index in this.matrix) {

            cells.push(this.matrix[index]);

        }

        return cells;

    }


}
