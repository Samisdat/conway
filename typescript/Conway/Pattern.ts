import {CellMatrix} from '@Conway/Conway/CellMatrix';
import {Position} from '@Conway/Conway/Position';

export class Pattern {

    private name: String;

    private matrix: CellMatrix;

    constructor(name: String, matrix: CellMatrix) {

        this.name = name;

        this.matrix = matrix;

    }

    public getMatrix(): CellMatrix {

        return this.matrix;
    }

    public getWidth(): number {
        return this.matrix.width();
    }

    public getHeight(): number {
        return this.matrix.height();
    }

    public mirrorVertical(): void {

        const mirror = new CellMatrix();

        for (const position of this.matrix.all()) {

            mirror.add(
                new Position(
                    position.x,
                    -1 * position.y
                )
            );

        }

        this.matrix = mirror;

    }

    public mirrorHorizontal(): void {

        const mirror = new CellMatrix();

        for (const position of this.matrix.all()) {

            mirror.add(
                new Position(
                    -1 * position.x,
                    position.y
                )
            );

        }

        this.matrix = mirror;

    }

}
