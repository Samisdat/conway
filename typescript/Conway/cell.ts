import { Position } from 'Conway/position';

export class Cell {

    public readonly position: Position;

    private nowliving: Boolean = false;

    private thenLiving: Boolean = true;

    constructor(position: Position) {
        this.position = position;
    }

    /**
     * Getter for x allows access like position.x
     */
    get x(): Number {
        return this.position.x;
    }

    get y(): Number {
        return this.position.y;
    }

    isAlive(): Boolean {
        return (true === this.nowliving);
    }

    kill(): void {
        this.thenLiving = false;
    }

    elapse(): void {

        if (true === this.thenLiving) {
            this.nowliving = true;
        }
        else {
            this.nowliving = false;
        }
    }

}