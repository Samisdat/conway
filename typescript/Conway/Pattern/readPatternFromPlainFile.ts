import {CellMatrix} from '@Conway/Conway/CellMatrix';
import {Position} from '@Conway/Conway/Position';
import {Pattern} from '@Conway/Conway/Pattern';

export const readPatternFromPlainFile = (patternPlainText: string): Pattern => {

    let name: string | undefined = undefined;
    const matrix = new CellMatrix();

    const patternPows: string[] = patternPlainText.split('\n');

    const isComment = new RegExp('^!(.*?)');
    const isMeta = new RegExp('^!(.*?)\:(.*?)$');

    let y = 0;
    let x = 0;

    for (let patternPow of patternPows) {

        patternPow = patternPow.trim();

        if (true === isComment.test(patternPow)) {

            if (true === isMeta.test(patternPow)) {

                const meta = isMeta.exec(patternPow);

                if (null === meta) {
                    continue;
                }

                if ('Name' === meta[1]) {
                    name = meta[2].trim();
                }

            }
            continue;
        }

        const cols: string[] = patternPow.trim().split('');

        x = 0;

        for (const col of cols) {

            if ('O' === col) {

                matrix.add(new Position(x, y));

            }

            x += 1;
        }

        y += 1;

    }

    if (undefined === name) {
        name = 'unnamed';
    }

    const pattern = new Pattern(name, matrix);
    //pattern.mirrorVertical();

    return pattern;

}

