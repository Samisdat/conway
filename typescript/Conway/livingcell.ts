import { Cell } from 'Conway/cell';
import { Position } from 'Conway/position';

export class LivingCell extends Cell {

    constructor(position: Position) {

        super(position);

        this.elapse();

    }

}