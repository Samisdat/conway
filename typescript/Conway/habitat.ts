import { Position } from './position';
import { Cell } from './cell';
import { LivingCell } from './livingcell';
import { Pattern } from './pattern';
import {Config} from '../Config';

export class Habitat {

    private generationDuration: number;

    private cells: Cell[] = [];

    constructor(generationDuration: number) {

        this.generationDuration = generationDuration;

    }

    public startAging(): void {

        setInterval(
            () => {
                this.elapse();
            },
            this.generationDuration
        );

    }

    private isLiving(position: Position): Boolean {

        let isLiving = false;

        for (let cell of this.cells) {

            if (cell.x === position.x && cell.y === position.y) {

                if (true === cell.isAlive()) {
                    isLiving = true;
                }

                break;
            }
        }

        return isLiving;
    }

    private countLivingNeighbours(position: Position): number {

        let livingNeighbours: number = 0;

        let neighbours: Position[] = position.getNeighbours();

        for (let neighbour of neighbours) {
            let isLiving = this.isLiving(neighbour);

            if (true === isLiving) {
                livingNeighbours += 1;
            }

        }

        return livingNeighbours;

    }

    private applyRules(): void {
        /**
         * there are four rules
         * 1) Any live cell with fewer than two live neighbours dies, as if caused by under-population.
         * 2) Any live cell with two or three live neighbours lives on to the next generation.
         * 3) Any live cell with more than three live neighbours dies, as if by over-population.
         * 4) Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
         *
         *  All rules dealing with living cells can be handled by kill all cells that don't have two or three neighbours
         *  Thats the first loop
         *
         *  To check if any death cells will be birth, we count the living neighbour of all cells neighbours
         *  Thats the second loop
         */

        for (let cell of this.cells) {

            let livingNeighbours = this.countLivingNeighbours(cell.position);

            if (2 === livingNeighbours || 3 === livingNeighbours) {
                continue;
            }

            cell.kill();

        }

        let createCellsAt = {};

        for (let cell of this.cells) {

            let neighbours = cell.position.getNeighbours();

            for (let neighbour of neighbours) {

                let livingNeighbours = this.countLivingNeighbours(neighbour);

                if (3 === livingNeighbours) {
                    createCellsAt[neighbour.x + '-' + neighbour.y] = neighbour;
                }

            }
        }

        for (let pos in createCellsAt) {
            let positionOfNewCell: Position = createCellsAt[pos];
            this.cells.push(new Cell(positionOfNewCell));
        }

    }

    private removeBodies(): void {

        let cleaned: Cell[] = [];

        for (let cell of this.cells) {
            if (true === cell.isAlive()) {
                cleaned.push(cell);
            }
        }

        this.cells = cleaned;

    }

    public elapse(): void {

        this.applyRules();

        for (let cell of this.cells) {
            cell.elapse();
        }

        this.removeBodies();

    }


    public seed(position: Position): void {
        this.cells.push(new LivingCell(position));
    }

    public seedPattern(pattern: Pattern): void {

        const positions = pattern.get();

        for (let position of positions) {
            this.cells.push(new LivingCell(position));
        }

    }

    public getAllCells(): Cell[] {

        return this.cells;

    }

    get(): Position[] {

        let living: Position[] = [];

        for (let cell of this.cells) {
            if (true === cell.isAlive()) {
                living.push(cell.position);
            }
        }

        return living;
    }

}