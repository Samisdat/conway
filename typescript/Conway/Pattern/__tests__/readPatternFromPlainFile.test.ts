import {readFileSync} from 'fs';
import serializer from '../../../../jest-serialize-conway';

import {readPatternFromPlainFile} from '@Conway/Conway/Pattern/readPatternFromPlainFile';
import {Pattern} from '@Conway/Conway/Pattern';

describe('PatternFromPlainText', () => {

    beforeAll(() => {
        expect.addSnapshotSerializer(serializer);
    });

    test('can be created with a complete *.cell file', () => {

        const gliderFile = readFileSync(__dirname + '/glider.cells', {encoding: 'utf8'});

        const gilder = readPatternFromPlainFile(gliderFile);

        expect(gilder).toBeInstanceOf(Pattern);

        expect(gilder.getMatrix()).toMatchSnapshot();

    });

    test('can be created with a .cell without meta ', () => {

        const patternStr = 'O..\n' +
            'OOO';

        const scholar = readPatternFromPlainFile(patternStr);

        expect(scholar).toBeInstanceOf(Pattern);

        expect(scholar.getMatrix()).toMatchSnapshot();

    });

});
