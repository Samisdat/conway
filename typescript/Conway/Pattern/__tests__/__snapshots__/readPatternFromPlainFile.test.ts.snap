// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PatternFromPlainText can be created with a .cell without meta  1`] = `
XXX
XOO
`;

exports[`PatternFromPlainText can be created with a complete *.cell file 1`] = `
XXX
OOX
OXO
`;
